File: frontend/static/widgets/free_form_content/remote_image.mjs
import { FreeFormContent } from './free_form_content.mjs'
import { ContentAndCaption } from '../containers/content_and_caption.mjs'
import { Caption } from '../caption.mjs'

/**
 * A piece of {@link FreeFormContent} which displays an image stored on a remote server.
 *
 * @augments FreeFormContent
 */
export class RemoteImage extends FreeFormContent {
  /**
   * @param {int} id the content's ID
   * @param {URL} src the image's URL
   * @param {?Caption} caption the image's caption
   */
  constructor ({ id, src, caption }) {
    super({ id })
    this.src = src
    this.caption = caption
  }

  /**
   * Deserialize the RemoteImage from its JSON API representation.
   *
   * @param obj
   * @returns {RemoteImage}
   */
  static fromJSON (obj) {
    return new RemoteImage({
      id: obj.id,
      src: obj.src,
      caption: Caption.maybeFromJSON(obj.caption)
    })
  }

  build () {
    const img = document.createElement('img')
    img.src = this.src

    return new ContentAndCaption({
      content: img,
      caption: this.caption
    })
  }
}

File: backend/src/server/free_form_content/remote_image.py
from datetime import datetime
from typing import Optional

from server.free_form_content.caption import Caption
from server.free_form_content.captioned_content import CaptionedContent
from server.util import combine


class RemoteImage(CaptionedContent):
    """An image post with its data stored on a remote server"""

    def __init__(
        self,
        src: str,
        caption: Optional[Caption],
        streams: list[int],
        content_id: Optional[int] = None,
        posted: Optional[datetime] = None,
    ):
        super().__init__(caption, streams, content_id, posted)
        self.src = src

    def type(self) -> str:
        return "remote_image"

    def to_db_json(self) -> dict:
        return combine(super().to_db_json(), {"src": self.src})
